sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -957,6 +946,36 @@ class Analyzer(override val catalogManager: CatalogManager)
           }
         }
     }
+
+    private def getMetadataAttributes(plan: LogicalPlan): Seq[Attribute] = {
+      lazy val childMetadataOutput = plan.children.flatMap(_.metadataOutput)
```

Comment:
nit: we can avoid building a new `Seq` frequently. The check can be 
`plan.children.exists(c => c.metadataOutput.exists(_.exprId == a.exprId))`

Reply:
The same to `hasMetadataCol`